libpromises/cf3lex.l

```diff
@@ -133,6 +138,57 @@ promise_type   [a-zA-Z_]+:
                           P.line_pos = 0;
                       }
 
+{macro_if}            {
+                        const char* version_text = yytext+18;
+                        ParserDebug("\tL:macro @if %d:version=%s\n", P.line_pos, version_text);
+                        {
+                          int request_major = 0;
+                          int request_minor = 0;
+                          int major = 0;
+                          int minor = 0;
+                          int patch = 0;
+                          if (3 == sscanf(Version(), "%d.%d.%d", &major, &minor, &patch))
+                          {
+                            if (2 == sscanf(version_text, "%d.%d", &request_major, &request_minor))
+                            {
+                              if (request_major > major || request_minor > minor)
+                              {
+                                /* ignore to the next @endif */
+                                ParserDebug("\tL:macro @if %d:ignoring to next @endif or EOF\n", P.line_pos);
+                                BEGIN(if_state);
+                              }
+                              else /* we're OK with the version */
+                              {
+                                ParserDebug("\tL:macro @if %d:accepted to next @endif or EOF\n", P.line_pos);
+                              }
+                            }
+                            else
+                            {
+                              yyerror("fatal: macro @if requested an unparseable version");
+                              exit(1);
+                            }
+                          }
+                          else
+                          {
+                            yyerror("fatal: Version() was unparseable\n");
+                          }
+                        }
+                      }
+
+{macro_endif}         {
+                        ParserDebug("\tL:macro @endif %d\n", P.line_pos);
+                        BEGIN(INITIAL);
+                      }
```

Comment:
All of the above fixed and tests added to ensure the stacking behavior is forbidden.

The EOF behavior when not in `if_ignored_state` was the only problem: I had to add `return 0` to avoid an endless loop.  I think I'm doing the right thing, just need confirmation.